test_plans/device-selector.asciidoc

```diff
@@ -0,0 +1,75 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for device selector
+
+This is a test plan for the device selector as described in SYCL 2020 Section 4.6.1.1.
+
+Estimated development time is three days.
+
+== Testing scope
+
+=== Backend coverage
+
+All the tests described below are not backend-specific and are performed for any SYCL backend.
+
+=== Device coverage
+
+All tests construct a test device for which conformance is assessed. All the tests described below are performed once for that test device.
+
+== Tests
+
+=== Predefined selector instances
+
+The file `device_selector/device_selector_predefined.cpp` currently implements tests for the SYCL 1.2.1. predefined device selectors `default_selector`, `gpu_selector`, `accelerator_selector`, and `cpu_selector`. These are marked as deprecated and duplicated to use the SYCL 2020 predefined device selectors `default_selector_v`, `gpu_selector_v`, `accelerator_selector_v`, and `cpu_selector_v` instead.
+
+Additionally when, for CPU, GPU, and accelerator, no device is available, it is checked whether constructing a device will throw an exception with the error code `errc::runtime`.
+
+=== Aspect selector
+
+==== Aspect coverage
+
+All aspects listed in the enum class `aspect` will be tested, let's identify these aspects as `{asp1, ..., aspN}`.
+
+The CTS will cover these combinations:
+
+- All combinations where no aspect is passed in.
+- All combinations where a single aspect is used.
+- All combinations of two aspects, including duplicates (e.g. `aspect_selector(aspX, aspX)`), and reversed order (e.g. `aspect_selector(aspX, aspY)` and `aspect_selector(aspY, aspX)`).
+
+The CTS does not have to test all combinations of 3 or more aspects.
+
+The CTS will test all the different numbers of aspects at least once, up to `N`, where the aspects are not duplicated, i.e. `aspect_selector(asp1), aspect_selector(asp1, asp2), ..., aspect_selector(asp1, asp2, ..., aspN)`.
+
+Additional testing will include 100 combinations of randomly selected aspect combinations, allowing aspect duplicates, with a random arity. The random combinations are selected at test generation time, with each test generation producing the same combinations and each run testing the same random combinations.
```

Comment:
This part of the test that chooses random combinations seems like overkill that we don't need for this first pass of the CTS, which focuses on "breadth first".  I'm not opposed to the random tests, but it seems lower priority to me.